### PR TITLE
Internal: Editor V2 - Separate build between packages and plugin [ED-10003]

### PR DESCRIPTION
**Pull request by Nevoss**

### CODE REVIEW

**Comment by Nevoss** [comment]
Depends on #21614

File: packages/packages/externalize-wordpress-assets-webpack-plugin/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+			return;
+		}
+
+		const { exact, startsWith } = this.options.externalsMap;
+
+		if ( Object.keys( exact ).includes( request ) ) {
+			return exact[ request ];
+		}
+
+		for ( const [ prefix, namespace ] of Object.entries( startsWith ) ) {
```
Comment (StyleShit) [inline packages/packages/externalize-wordpress-assets-webpack-plugin/src/index.ts:70]: Shouldn't it be one of those?

```suggestion
		for ( const [ prefix, referenceKey ] of Object.entries( startsWith ) ) {
```

```suggestion
		for ( const [ prefix, globalKey ] of Object.entries( startsWith ) ) {
```

File: packages/packages/generate-wordpress-asset-file-webpack-plugin/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	transformIntoDepName( name: string ) {
+		const { startsWith, exact } = this.options.handlesMap;
+
+		if ( Object.keys( exact ).includes( name ) ) {
+			return exact[ name ];
+		}
+
+		for ( const [ key, value ] of Object.entries( startsWith ) ) {
+			if ( name.startsWith( key ) ) {
+				return name.replace( key, value );
```
Comment (StyleShit) [inline packages/packages/generate-wordpress-asset-file-webpack-plugin/src/index.ts:154]: Probably nit-picking, but it should be something like this IMO:
```suggestion
				return name.replace( new RegExp('^' + key, 'i' ), value );
```

File: packages/packages/generate-wordpress-asset-file-webpack-plugin/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	isExternalDep( request: string ) {
+		const { startsWith, exact } = this.options.handlesMap;
+
+		return request && (
+			Object.keys( exact ).includes( request ) ||
+			Object.keys( startsWith ).some( ( dep ) => request.startsWith( dep ) )
+		);
+	}
+
+	transformIntoDepName( name: string ) {
```
Comment (StyleShit) [inline packages/packages/generate-wordpress-asset-file-webpack-plugin/src/index.ts:145]:
```suggestion
	getHandleFromDep( dep: string ) {
```
Maybe?
also, let's talk about it

**Review by StyleShit — CHANGES_REQUESTED** [review]

**Comment by StyleShit** [comment]
copilot:poem

File: .grunt-config/webpack.js [hunk 4]
```diff
@@ -281,6 +284,9 @@ webpackProductionConfig.forEach( ( config, index ) => {
 	}
 } );
 
+// The 'packages' config not needed a .min suffix (it has it's own config).
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply .grunt-config/webpack.js:287]: Done

File: .grunt-config/webpack.packages.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
+const path = require("path");
+const fs = require("fs");
+const { GenerateWordPressAssetFileWebpackPlugin } = require("@elementor/generate-wordpress-asset-file-webpack-plugin");
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply .grunt-config/webpack.packages.js:3]: done

File: .grunt-config/webpack.packages.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	} );
+
+const common = {
+	name: 'packages',
+	entry: Object.fromEntries(
+		packages.map( ( { name, path } ) => [ name, path ] )
+	),
+	plugins: [
+		new GenerateWordPressAssetFileWebpackPlugin( { handlePrefix: 'elementor-packages-' } ),
+		new ExternalizeWordPressAssetsWebpackPlugin( { referenceKey: '__UNSTABLE__elementorPackages' } ),
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply .grunt-config/webpack.packages.js:37]: Done

File: .grunt-config/webpack.packages.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+	],
+	output: {
+		path: path.resolve( __dirname, '../assets/js/packages/' ),
+	},
+}
+
+const devConfig = {
+	...common,
+	mode: 'development',
+	devtool: false, // TODO: Need to check what to do with source maps.
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply .grunt-config/webpack.packages.js:47]: Done

File: .grunt-config/webpack.packages.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+	watch: true, // All the webpack config in the plugin that are dev, should have this property.
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply .grunt-config/webpack.packages.js:48]: Seems weird to me that the dev build should have `watch` by default and you don't pass `--watch` flag

File: .grunt-config/webpack.packages.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+	output: {
+		...( common.output || {} ),
+		filename: '[name].js',
+	}
+}
+
+const prodConfig = {
+	...common,
+	mode: 'production',
+	devtool: false, // TODO: Need to check what to do with source maps.
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply .grunt-config/webpack.packages.js:58]: Yeah, by default it generate sperate file

File: package.json [hunk 10]
```diff
@@ -49,6 +49,9 @@
         "@babel/preset-env": "~7.18.6",
         "@babel/runtime-corejs3": "~7.18.6",
         "@elementor/eslint-plugin-editor": "^0.4.0",
+        "@elementor/generate-wordpress-asset-file-webpack-plugin": "file:./packages/packages/generate-wordpress-asset-file-webpack-plugin",
+        "@elementor/extract-i18n-expressions-webpack-plugin": "file:./packages/packages/extract-i18n-expressions-webpack-plugin",
+        "@elementor/externalize-wordpress-assets-webpack-plugin": "file:./packages/packages/externalize-wordpress-assets-webpack-plugin",
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply package.json:54]: Done

File: packages/.eslintignore [hunk 11]
```diff
@@ -1 +1,2 @@
 node_modules/
+packages/*/dist/
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply packages/.eslintignore:2]: Not required but done

File: packages/.gitignore [hunk 12]
```diff
@@ -1,3 +1,4 @@
+packages/*/dist/
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply packages/.gitignore:1]: Done

File: packages/packages/externalize-wordpress-assets-webpack-plugin/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,116 @@
+import { ExternalsPlugin, Compiler } from 'webpack';
+
+type ExternalsMap = {
+	exact: Record< string, string >;
+	startsWith: Record< string, string >;
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply packages/packages/externalize-wordpress-assets-webpack-plugin/src/index.ts:5]: Done

File: packages/packages/externalize-wordpress-assets-webpack-plugin/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+	referenceKey: string,
+	externalsMap: ExternalsMap
+}
+
+type Options = {
+	referenceKey: string,
+	externalsMap?: Partial<ExternalsMap>
+}
+
+const baseExternalsMap: ExternalsMap = {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply packages/packages/externalize-wordpress-assets-webpack-plugin/src/index.ts:18]: This is not the default, the is the base and then the options extends it

File: packages/packages/externalize-wordpress-assets-webpack-plugin/src/index.ts (hunk 1 above)
Comment (Nevoss) [reply packages/packages/externalize-wordpress-assets-webpack-plugin/src/index.ts:70]: Done

File: packages/packages/generate-wordpress-asset-file-webpack-plugin/src/index.ts (hunk 2 above)
Comment (Nevoss) [reply packages/packages/generate-wordpress-asset-file-webpack-plugin/src/index.ts:154]: Not sure about it

File: packages/packages/generate-wordpress-asset-file-webpack-plugin/src/index.ts (hunk 3 above)
Comment (Nevoss) [reply packages/packages/generate-wordpress-asset-file-webpack-plugin/src/index.ts:145]: Done

**Pull request closed without merging**